### ISSUE TITLE
contrastive is inconsistency between papers and code

### ISSUE THREAD

**Issue by pickwu** [amazon-research/sccl#27]
In section3.1 of  the paper  , contrastive loss = -log(pos/neg).but in code,contrastive loss = -log(pos/(neg+pos)).Why the denominator needs to add pos in the code?

**Comment by pickwu** [comment]
`class PairConLoss(nn.Module):
    def __init__(self, temperature=0.05):
        super(PairConLoss, self).__init__()
        self.temperature = temperature
        self.eps = 1e-08
        print(f"\n Initializing PairConLoss \n")

    def forward(self, features_1, features_2):
        device = features_1.device
        batch_size = features_1.shape[0]
        features= torch.cat([features_1, features_2], dim=0) 
        mask = torch.eye(batch_size, dtype=torch.bool).to(device) 
        mask = mask.repeat(2, 2) 
        mask = ~mask
        
        pos = torch.exp(torch.sum(features_1*features_2, dim=-1) / self.temperature) 
        pos = torch.cat([pos, pos], dim=0) 

        neg = torch.exp(torch.mm(features, features.t().contiguous()) / self.temperature) 
        neg = neg.masked_select(mask).view(2*batch_size, -1) 
        
        neg_mean = torch.mean(neg)
        pos_n = torch.mean(pos)
        Ng = neg.sum(dim=-1) # 40
            
        loss_pos = (- torch.log(pos / (Ng+pos))).mean() 
        
        return {"loss":loss_pos, "pos_mean":pos_n.detach().cpu().numpy(), "neg_mean":neg_mean.detach().cpu().numpy(), "pos":pos.detach().cpu().numpy(), "neg":neg.detach().cpu().numpy()}`

**Comment by pickwu** [comment]
`class PairConLoss(nn.Module):
……
def forward(self, features_1, features_2):
    device = features_1.device
    batch_size = features_1.shape[0]
    features= torch.cat([features_1, features_2], dim=0) 
    mask = torch.eye(batch_size, dtype=torch.bool).to(device) 
    mask = mask.repeat(2, 2) 
    mask = ~mask
    
    pos = torch.exp(torch.sum(features_1*features_2, dim=-1) / self.temperature) 
    pos = torch.cat([pos, pos], dim=0) 

    neg = torch.exp(torch.mm(features, features.t().contiguous()) / self.temperature) 
    neg = neg.masked_select(mask).view(2*batch_size, -1) 
    
    neg_mean = torch.mean(neg)
    pos_n = torch.mean(pos)
    Ng = neg.sum(dim=-1) # 40
        
    loss_pos = (- torch.log(pos / (Ng+pos))).mean() 
    
    return {"loss":loss_pos, "pos_mean":pos_n.detach().cpu().numpy(), "neg_mean":neg_mean.detach().cpu().numpy(), "pos":pos.detach().cpu().numpy(), "neg":neg.detach().cpu().numpy()}``

**Comment by pickwu** [comment]
`class PairConLoss(nn.Module):
……    
    loss_pos = (- torch.log(pos / (Ng+pos))).mean() 
    return {"loss":loss_pos, "pos_mean":pos_n.detach().cpu().numpy(), "neg_mean":neg_mean.detach().cpu().numpy(), "pos":pos.detach().cpu().numpy(), "neg":neg.detach().cpu().numpy()}`

**Comment by pickwu** [comment]
`
class PairConLoss(nn.Module):
……    
    loss_pos = (- torch.log(pos / (Ng+pos))).mean() 
    return {"loss":loss_pos, "pos_mean":pos_n.detach().cpu().numpy(), "neg_mean":neg_mean.detach().cpu().numpy(), "pos":pos.detach().cpu().numpy(), "neg":neg.detach().cpu().numpy()}
`

**Comment by pickwu** [comment]
> class PairConLoss(nn.Module):
……    
    loss_pos = (- torch.log(pos / (Ng+pos))).mean() 
    return {"loss":loss_pos, "pos_mean":pos_n.detach().cpu().numpy(), "neg_mean":neg_mean.detach().cpu().numpy(), "pos":pos.detach().cpu().numpy(), "neg":neg.detach().cpu().numpy()}
